### PR TITLE
Allow obtaining Inference API tokens with whoami

### PR DESCRIPTION
**Pull request by osanseviero**
This adds an optional argument to the existing `whoami` to obtain the full response (which uses `whoami-v2`)  to get more info. This includes the Inference API token which is needed for the wrapper client.

### CODE REVIEW

**Comment by julien-c** [comment]
looks good from a quick glance but let's wait for @Pierrci and @LysandreJik feedback

File: tests/test_hf_api.py
Comment (LysandreJik) [inline tests/test_hf_api.py:93]:
```suggestion
```

File: src/huggingface_hub/hf_api.py
Comment (LysandreJik) [inline src/huggingface_hub/hf_api.py:180]:
```suggestion
        self, token: str, full: Optional[bool] = None
```

**Review by LysandreJik — COMMENTED** [review]
This looks like a great addition! I'm not too sure of the return type being a tuple in one case and a dict in another case. I'd rather we always return a dict and break the API.

Do you know of any library that has not pinned `hfh` to a specific version and is using `whoami`, that would be impacted by this change?

The `whoami-v2` endpoint seems to be deployed only for HF, let's not forget to deploy it globally before merging this.

**Comment by julien-c** [comment]
depending on feedback from @Pierrci we might even just remove the whoami (non-full) option, especially if we're breaking BC here anyways

**Comment by osanseviero** [comment]
Yeah, I kept them separate to avoid breaking people. If we're ok breaking the API since no libraries are using this method, then let's go with the full method.

**Comment by osanseviero** [comment]
@Pierrci ping. Are we ok going with whoami-v2 all the time?

**Comment by Pierrci** [comment]
What I would really prefer is to never have exposed `whoami-v2` at all x)

But since it's already used elsewhere, I don't want to discriminate against hfhub; and whether all the time or not won't make a big difference, so fine with the former option :)

**Comment by osanseviero** [comment]
> What I would really prefer is to never have exposed whoami-v2 at all x)
We want to expose this because we need the inference API tokens for the inference client, so I think it's nice to allow them to obtain this programmatically.

**Comment by osanseviero** [comment]
@LysandreJik I changed this to always use whoami-v2.


> The whoami-v2 endpoint seems to be deployed only for HF, let's not forget to deploy it globally before merging this.

@Pierrci is this still the case?

**Comment by Pierrci** [comment]
> We want to expose this because we need the inference API tokens for the inference client, so I think it's nice to allow them to obtain this programmatically.

Oh yes yes, I'm not criticizing what you're trying to achieve haha, it was more for myself :)

> The whoami-v2 endpoint seems to be deployed only for HF, let's not forget to deploy it globally before merging this.

What makes you say that @LysandreJik? I don't think anything in the code is limiting the use of `whoami-v2` to HF :) (it's already used in [`autonlp`](https://github.com/huggingface/autonlp/blob/4f47c96e5d6896990d18c2b317c6f9b2c99a5e00/src/autonlp/autonlp.py#L43) without problems I think cc @SBrandeis)

**Comment by LysandreJik** [comment]
When trying it out last week it errored out on my `Jikiwa` profile but not on my `Lysandre` profile ... it must have been user error and I incorrectly assumed it was because it was HF-only :) just tried it and it works wonders

File: src/huggingface_hub/commands/user.py
Comment (LysandreJik) [inline src/huggingface_hub/commands/user.py:161]:
```suggestion
```

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @osanseviero!

**Comment by osanseviero** [comment]
> When trying it out last week it errored out on my Jikiwa profile but not on my Lysandre profile ... it must have been user error and I incorrectly assumed it was because it was HF-only :) just tried it and it works wonders

Yeah, I'm seeing a funny behavior with the tests. `test_hf_api.py` passes (and uses `whoami-v2`) but `test_repository.py` is giving me ` Client Error: Unauthorized for url: https://huggingface.co/api/whoami-v2`

File: src/huggingface_hub/commands/user.py
Comment (osanseviero) [reply src/huggingface_hub/commands/user.py:161]: Why? The `print` is wanted here, this is the CLI.

**Comment by osanseviero** [comment]
It seems like I algo get the `Unauthorized for url` error with `whoami` (v1(, so that's unrelated to the PR.

I'll merge in few hours unless there are objections 👨‍⚖️

**Pull request merged**
